iree/compiler/Dialect/Flow/IR/FlowTypes.h

```diff
@@ -16,14 +16,207 @@
 #define IREE_COMPILER_DIALECT_FLOW_IR_FLOWTYPES_H_
 
 #include "iree/compiler/Dialect/IREE/IR/IREETypes.h"
+#include "iree/compiler/Dialect/Shape/IR/ShapeTypes.h"
 #include "llvm/ADT/DenseMapInfo.h"
 #include "llvm/ADT/SmallVector.h"
 #include "llvm/ADT/StringSwitch.h"
+#include "mlir/IR/BuiltinTypes.h"
+#include "mlir/IR/DialectImplementation.h"
+#include "mlir/IR/OpImplementation.h"
 #include "mlir/IR/TypeSupport.h"
 #include "mlir/IR/Types.h"
 #include "mlir/Support/LLVM.h"
 
 // Order matters.
 #include "iree/compiler/Dialect/Flow/IR/FlowEnums.h.inc"
 
+namespace mlir {
+namespace iree_compiler {
+namespace IREE {
+namespace Flow {
+
+//===----------------------------------------------------------------------===//
+// Object types
+//===----------------------------------------------------------------------===//
+
+namespace detail {
+struct DispatchTensorTypeStorage;
+}  // namespace detail
+
+// Blatantly ripped from ShapedType, because the closed type system means that
```

Comment:
This has different semantics - these are representative of external storage that can be mutated and don't follow tensor semantics - that's how we can support loading/storing arbitrary tiles of the tensors  regardless of which workgroup is being processed. If we used tensor then we would still need special ops (`flow.pretend_tensors_are_actually_stateful_storage_and_store_into_them`) and risk that all kinds of passes won't obey that as tensors are assumed immutable.

Since upstream doesn't want a nice way to represent shapes I was chatting with Stella and in the future we'll likely just kill the shapex dialect and move ranked_shape into the flow dialect - in which case this will not feel like duplication :)

Reply:
Ok. this is better typed. THanks!